src/generator/qbe.rs

```diff
@@ -0,0 +1,418 @@
+                let tmp = self.make_tmp();
+
+                match op {
+                    BinOp::Addition => {
+                        self.push_assign_instr(
+                            &tmp,
+                            // TODO: should not assume this is always word
+                            QbeType::Word,
+                            Instr::Add(lhs, rhs),
+                        );
+                    }
+                    BinOp::Subtraction => {
+                        self.push_assign_instr(&tmp, QbeType::Word, Instr::Sub(lhs, rhs));
+                    }
+                    _ => todo!("binop: {:?}", op),
+                }
+
+                Ok(tmp)
+            }
+            _ => todo!("expression: {:?}", expr),
+        }
+    }
+
+    /// Returns a QBE type for AST type
+    fn get_type(&mut self, ty: &Type) -> GenerateResult<QbeType> {
+        match ty {
+            Type::Any => Err("'any' type is not supported".into()),
+            Type::Int => Ok(QbeType::Word),
+            Type::Bool => Ok(QbeType::Byte),
+            Type::Str | Type::Array(_, _) | Type::Struct(_) => todo!("aggregate types"),
+        }
+    }
+
+    /// Returns a new temporary assigned to a variable
+    fn make_var(&mut self, name: &str) -> GenerateResult<Temporary> {
+        // XXX: should shadowing be allowed?
+        if self.lookup_var(name).is_ok() {
+            return Err(format!("Re-declaration of variable '{}'", name));
+        }
+
+        let tmp = self.make_tmp();
+
+        let scope = self
+            .scopes
+            .last_mut()
+            .expect("Expected top scope to be present");
+        scope.insert(name.to_owned(), tmp.clone());
+
+        Ok(tmp)
+    }
+
+    /// Returns a temporary corresponding to the given variable
+    fn lookup_var(&self, name: &str) -> GenerateResult<&Temporary> {
+        self.scopes
+            .iter()
+            .rev()
+            .filter_map(|scope| scope.get(name))
+            .next()
+            .ok_or_else(|| format!("Undefined variable '{}'", &name))
+    }
+
+    /// Returns a new unique temporary
+    fn make_tmp(&mut self) -> Temporary {
+        self.tmp_counter += 1;
+        Temporary::new(format!("temp{}", self.tmp_counter))
+    }
+
+    /// Appends text to the buffer
+    fn push(&mut self, s: &str) {
+        self.buf.push_str(s);
+    }
+
+    /// Appends a @label
+    fn push_label(&mut self, label: &Label) {
+        self.push(&format!("{}\n", label));
+    }
+
+    /// Appends an instruction
+    fn push_instr(&mut self, instr: Instr) {
+        self.push(&format!("\t{}\n", instr));
+    }
+
+    /// Appends an instruction assigned to a temporary
+    fn push_assign_instr(&mut self, tmp: &Temporary, ty: QbeType, instr: Instr) {
+        self.push(&format!("\t{} ={} {}\n", tmp, ty, instr));
+    }
+}
+
+use std::fmt;
+
+/// QBE temporary
+#[derive(Debug, Clone)]
+pub(crate) struct Temporary {
```

Comment:
Can these stay inside the generator? 

```suggestion
pub(super) struct Temporary {
```

(Haven't checked if this compiles)

Reply:
If you want to stay inside a generator, you need to do `pub(self) ...`, `super` refers to the parent crate i.e. `crate::generator`